pkg/instrumentation/nethttp/server/server_hook.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/nethttp/semconv"
+	"github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/shared"
+)
+
+const (
+	instrumentationName    = "github.com/open-telemetry/opentelemetry-go-compile-instrumentation/pkg/instrumentation/nethttp"
+	instrumentationVersion = "0.1.0"
+)
+
+var (
+	logger     = shared.GetLogger()
+	tracer     trace.Tracer
+	propagator propagation.TextMapPropagator
+	initOnce   sync.Once
+)
+
+func initInstrumentation() {
+	initOnce.Do(func() {
+		if err := shared.SetupOTelSDK(); err != nil {
+			logger.Error("failed to setup OTel SDK", "error", err)
+		}
+		tracer = otel.GetTracerProvider().Tracer(
+			instrumentationName,
+			trace.WithInstrumentationVersion(instrumentationVersion),
+		)
+		propagator = otel.GetTextMapPropagator()
+		logger.Info("HTTP server instrumentation initialized")
+	})
+}
+
+// netHttpServerEnabler controls whether server instrumentation is enabled
+type netHttpServerEnabler struct{}
+
+func (n netHttpServerEnabler) Enable() bool {
+	return shared.Instrumented("NETHTTP")
+}
+
+var serverEnabler = netHttpServerEnabler{}
+
+func BeforeServeHTTP(ictx inst.HookContext, recv interface{}, w http.ResponseWriter, r *http.Request) {
+	if !serverEnabler.Enable() {
+		logger.Debug("HTTP server instrumentation disabled")
+		return
+	}
+
+	initInstrumentation()
+
+	logger.Debug("BeforeServeHTTP called",
+		"method", r.Method,
+		"url", r.URL.String(),
+		"remote_addr", r.RemoteAddr)
+
+	// Extract trace context from incoming request headers
+	ctx := propagator.Extract(r.Context(), propagation.HeaderCarrier(r.Header))
+
+	// Get trace attributes from semconv
+	attrs := semconv.HTTPServerRequestTraceAttrs("", r)
+
+	// Get HTTP route from r.Pattern (Go 1.22+)
+	route := semconv.HTTPRoute(r.Pattern)
+	spanName := semconv.HTTPServerSpanName(r.Method, route)
+
+	// Start span
+	ctx, span := tracer.Start(ctx,
+		spanName,
+		trace.WithSpanKind(trace.SpanKindServer),
+		trace.WithAttributes(attrs...),
+	)
+
+	// Add route attribute if available
+	if route != "" {
+		span.SetAttributes(semconv.HTTPServerRoute(route))
+	}
+
+	// Wrap ResponseWriter to capture status code
+	wrapper := &writerWrapper{
+		ResponseWriter: w,
+		statusCode:     http.StatusOK,
+	}
+	ictx.SetParam(1, wrapper)
+
+	// Store data for after hook
+	ictx.SetData(map[string]interface{}{
+		"ctx":   ctx,
+		"span":  span,
+		"start": time.Now(),
+	})
+}
+
+func AfterServeHTTP(ictx inst.HookContext) {
+	if !serverEnabler.Enable() {
+		return
+	}
+
+	data, ok := ictx.GetData().(map[string]interface{})
+	if !ok || data == nil {
+		logger.Debug("AfterServeHTTP: no data from before hook")
+		return
+	}
+
+	span, ok := data["span"].(trace.Span)
+	if !ok || span == nil {
+		logger.Debug("AfterServeHTTP: no span from before hook")
+		return
+	}
+	defer span.End()
+
+	startTime, _ := data["start"].(time.Time)
```

Comment:
nitpick: I think `startTime` closer to its usage, just before `logger.Debug` call, would be better